src/PowerShellEditorServices/Services/DebugAdapter/BreakpointSyncService.cs

```diff
@@ -0,0 +1,904 @@
+        int Line = 0,
+        int Column = 0);
+
+    internal record struct BreakpointHandle(SemaphoreSlim Target) : IDisposable
+    {
+        public void Dispose() => Target.Release();
+    }
+
+    private readonly BreakpointMap _map = new(new(), new());
+
+    private readonly ConditionalWeakTable<Runspace, BreakpointMap> _attachedRunspaceMap = new();
+
+    private readonly ConcurrentBag<SyncedBreakpoint> _pendingAdds = new();
+
+    private readonly ConcurrentBag<SyncedBreakpoint> _pendingRemovals = new();
+
+    private readonly SemaphoreSlim _breakpointMutateHandle = AsyncUtils.CreateSimpleLockingSemaphore();
+
+    private readonly IInternalPowerShellExecutionService _host;
+
+    private readonly ILogger<BreakpointSyncService> _logger;
+
+    private readonly DebugStateService _debugStateService;
+
+    private readonly ILanguageServerFacade _languageServer;
+
+    private bool? _isSupported;
+
+    public BreakpointSyncService(
+        IInternalPowerShellExecutionService host,
+        ILoggerFactory loggerFactory,
+        DebugStateService debugStateService,
+        ILanguageServerFacade languageServer)
+    {
+        _host = host;
+        _logger = loggerFactory.CreateLogger<BreakpointSyncService>();
+        _debugStateService = debugStateService;
+        _languageServer = languageServer;
+    }
+
+    public bool IsSupported
+    {
+        get => _isSupported ?? false;
+        set
+        {
+            Debug.Assert(_isSupported is null, "BreakpointSyncService.IsSupported should only be set once by initialize.");
+            _isSupported = value;
+        }
+    }
+
+    public bool IsMutatingBreakpoints
+    {
+        get
+        {
+            if (_breakpointMutateHandle.Wait(0))
+            {
+                _breakpointMutateHandle.Release();
+                return false;
+            }
+
+            return true;
+        }
+    }
+
+    private int? TargetRunspaceId => _debugStateService?.RunspaceId;
+
+    public async Task UpdatedByServerAsync(BreakpointUpdatedEventArgs eventArgs)
+    {
+        if (!_map.ByPwshId.TryGetValue(eventArgs.Breakpoint.Id, out SyncedBreakpoint existing))
+        {
+            // If we haven't told the client about the breakpoint yet then we can just ignore.
+            if (eventArgs.UpdateType is BreakpointUpdateType.Removed)
+            {
+                return;
+            }
+
+            existing = CreateFromServerBreakpoint(eventArgs.Breakpoint);
+            string? id = await SendToClientAsync(existing.Client, BreakpointUpdateType.Set)
+                .ConfigureAwait(false);
+
+            existing.Client.Id = id!;
+            RegisterBreakpoint(existing);
+            return;
+        }
+
+        if (eventArgs.UpdateType is BreakpointUpdateType.Removed)
+        {
+            UnregisterBreakpoint(existing);
+            await SendToClientAsync(existing.Client, BreakpointUpdateType.Removed).ConfigureAwait(false);
+            return;
+        }
```

Comment:
Wait does the `out` param in the `if` statement above _really_ declare it in the scope above the conditional?! That's confusing scope. Maybe my confusion would be better alleviated by doing the `TryGet` first, and then check the return result only in the `if`, so that the `out` param looks like it's declared in the scope it is actually declared in.